File: android/cross-promo/src/main/kotlin/kr/dev/hoony/crosspromo/data/FirestoreService.kt
package kr.dev.hoony.crosspromo.data

import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import kr.dev.hoony.crosspromo.model.PromoApp
import kr.dev.hoony.crosspromo.model.PromoConfig
import org.json.JSONObject
import java.net.HttpURLConnection
import java.net.URL

internal class FirestoreService(private val projectId: String) {

    private val baseUrl =
        "https://firestore.googleapis.com/v1/projects/$projectId/databases/(default)/documents"

    suspend fun fetchApps(): List<PromoApp> = withContext(Dispatchers.IO) {
        val json = httpGet("$baseUrl/cross_promo_apps")
        val documents = json.optJSONArray("documents") ?: return@withContext emptyList()

        (0 until documents.length()).mapNotNull { i ->
            try {
                val doc = documents.getJSONObject(i)
                val fields = doc.getJSONObject("fields")
                parsePromoApp(fields)
            } catch (e: Exception) {
                null
            }
        }
    }

    suspend fun fetchConfig(): PromoConfig = withContext(Dispatchers.IO) {
        try {
            val json = httpGet("$baseUrl/cross_promo_config/settings")
            val fields = json.getJSONObject("fields")
            PromoConfig(
                enabled = fields.optJSONObject("enabled")?.optBoolean("booleanValue") ?: true,
                selectionMode = fields.optJSONObject("selectionMode")?.optString("stringValue")
                    ?: "weighted_random",
                minAppOpenCount = fields.optJSONObject("minAppOpenCount")
                    ?.optString("integerValue")?.toIntOrNull() ?: 3,
                popupTitle = fields.optJSONObject("popupTitle")?.optString("stringValue")
                    ?: "개발자의 다른 앱",
            )
        } catch (e: Exception) {
            PromoConfig()
        }
    }

    private fun parsePromoApp(fields: JSONObject): PromoApp? {
        val appId = fields.optJSONObject("appId")?.optString("stringValue") ?: return null

        val platforms = mutableListOf<String>()
        fields.optJSONObject("platforms")
            ?.optJSONObject("arrayValue")
            ?.optJSONArray("values")
            ?.let { arr ->
                for (j in 0 until arr.length()) {
                    arr.getJSONObject(j).optString("stringValue")?.let { platforms.add(it) }
                }
            }

        val storeLinks = mutableMapOf<String, String>()
        fields.optJSONObject("storeLinks")
            ?.optJSONObject("mapValue")
            ?.optJSONObject("fields")
            ?.let { linksFields ->
                linksFields.keys().forEach { key ->
                    linksFields.optJSONObject(key)?.optString("stringValue")?.let {
                        storeLinks[key] = it
                    }
                }
            }

        return PromoApp(
            appId = appId,
            appName = fields.optJSONObject("appName")?.optString("stringValue") ?: "",
            shortDescription = fields.optJSONObject("shortDescription")?.optString("stringValue")
                ?: "",
            iconUrl = fields.optJSONObject("iconUrl")?.optString("stringValue") ?: "",
            platforms = platforms,
            storeLinks = storeLinks,
            iosAppStoreId = fields.optJSONObject("iosAppStoreId")?.optString("stringValue") ?: "",
            priority = fields.optJSONObject("priority")?.optString("integerValue")?.toIntOrNull()
                ?: 10,
            enabled = fields.optJSONObject("enabled")?.optBoolean("booleanValue") ?: true,
        )
    }

    private fun httpGet(urlString: String): JSONObject {
        val url = URL(urlString)
        val connection = url.openConnection() as HttpURLConnection
        connection.requestMethod = "GET"
        connection.connectTimeout = 10_000
        connection.readTimeout = 10_000

        return try {
            val responseCode = connection.responseCode
            if (responseCode == HttpURLConnection.HTTP_OK) {
                val body = connection.inputStream.bufferedReader().readText()
                JSONObject(body)
            } else {
                throw Exception("HTTP $responseCode")
            }
        } finally {
            connection.disconnect()
        }
    }
}

File: ios/Sources/CrossPromoSDK/Data/FirestoreService.swift
import Foundation

final class FirestoreService {

    private let baseUrl: String

    init(projectId: String = "crosspromosdk") {
        self.baseUrl = "https://firestore.googleapis.com/v1/projects/\(projectId)/databases/(default)/documents"
    }

    func fetchApps() async throws -> [PromoApp] {
        let data = try await httpGet("\(baseUrl)/cross_promo_apps")
        let json = try JSONSerialization.jsonObject(with: data) as? [String: Any] ?? [:]
        guard let documents = json["documents"] as? [[String: Any]] else {
            return []
        }

        return documents.compactMap { doc -> PromoApp? in
            guard let fields = doc["fields"] as? [String: Any] else { return nil }
            return parsePromoApp(fields)
        }
    }

    func fetchConfig() async throws -> PromoConfig {
        let data = try await httpGet("\(baseUrl)/cross_promo_config/settings")
        let json = try JSONSerialization.jsonObject(with: data) as? [String: Any] ?? [:]
        guard let fields = json["fields"] as? [String: Any] else {
            return PromoConfig()
        }

        return PromoConfig(
            enabled: boolValue(fields, "enabled") ?? true,
            selectionMode: stringValue(fields, "selectionMode") ?? "weighted_random",
            minAppOpenCount: intValue(fields, "minAppOpenCount") ?? 3,
            popupTitle: stringValue(fields, "popupTitle") ?? "개발자의 다른 앱"
        )
    }

    // MARK: - Firestore REST API Parsing

    private func parsePromoApp(_ fields: [String: Any]) -> PromoApp? {
        guard let appId = stringValue(fields, "appId") else { return nil }

        var platforms: [String] = []
        if let arrayValue = (fields["platforms"] as? [String: Any])?["arrayValue"] as? [String: Any],
           let values = arrayValue["values"] as? [[String: Any]] {
            platforms = values.compactMap { $0["stringValue"] as? String }
        }

        var storeLinks: [String: String] = [:]
        if let mapValue = (fields["storeLinks"] as? [String: Any])?["mapValue"] as? [String: Any],
           let linksFields = mapValue["fields"] as? [String: Any] {
            for (key, value) in linksFields {
                if let sv = (value as? [String: Any])?["stringValue"] as? String {
                    storeLinks[key] = sv
                }
            }
        }

        return PromoApp(
            appId: appId,
            appName: stringValue(fields, "appName") ?? "",
            shortDescription: stringValue(fields, "shortDescription") ?? "",
            iconUrl: stringValue(fields, "iconUrl") ?? "",
            platforms: platforms,
            storeLinks: storeLinks,
            iosAppStoreId: stringValue(fields, "iosAppStoreId") ?? "",
            priority: intValue(fields, "priority") ?? 10,
            enabled: boolValue(fields, "enabled") ?? true
        )
    }

    private func stringValue(_ fields: [String: Any], _ key: String) -> String? {
        (fields[key] as? [String: Any])?["stringValue"] as? String
    }

    private func intValue(_ fields: [String: Any], _ key: String) -> Int? {
        guard let str = (fields[key] as? [String: Any])?["integerValue"] as? String else { return nil }
        return Int(str)
    }

    private func boolValue(_ fields: [String: Any], _ key: String) -> Bool? {
        (fields[key] as? [String: Any])?["booleanValue"] as? Bool
    }

    // MARK: - HTTP

    private func httpGet(_ urlString: String) async throws -> Data {
        guard let url = URL(string: urlString) else {
            throw URLError(.badURL)
        }
        var request = URLRequest(url: url)
        request.httpMethod = "GET"
        request.timeoutInterval = 10
        let (data, response) = try await URLSession.shared.data(for: request)
        guard let httpResponse = response as? HTTPURLResponse,
              httpResponse.statusCode == 200 else {
            throw URLError(.badServerResponse)
        }
        return data
    }
}
